### PR TITLE
sql/catalog: pull virtual schemas under Collect, remove `catalog/accessors`, cleanup

### PR DESCRIPTION
**Pull request by ajwerner**
The thrust of this PR is to remove the `LogicalSchemaAccessor` concept and pull "precise" descriptor resolution underneath the `descs.Collection`. Along the way, this PR performs some cleanup of the interfaces and the descs.Collection. This is a small part of a much broader effort to isolate and clean up the entire concept of descriptor resolution.

Release note: None

### CODE REVIEW

File: pkg/sql/catalog/descs/collection.go [hunk 1]
```diff
@@ -1937,6 +2051,31 @@ func (tc *Collection) GetObjectNamesAndIDs(
 	return tableNames, tableIDs, nil
 }
 
+func (tc *Collection) maybeGetVirtualObjectNamesAndIDs(
+	scName string, dbDesc catalog.DatabaseDescriptor, flags tree.DatabaseListFlags,
+) (ok bool, _ tree.TableNames, _ descpb.IDs, _ error) {
```
Comment (postamar) [inline pkg/sql/catalog/descs/collection.go:2056]: This function never returns an error.

File: pkg/sql/catalog/descs/collection.go [hunk 2]
```diff
@@ -481,19 +500,89 @@ func (tc *Collection) getDatabaseByName(
...
+	); isVirtual || err != nil {
+		return desc, err
+	}
+	// Resolve type aliases which are usually available in the PostgreSQL as an extension
+	// on the public schema.
+	// TODO(ajwerner): Pull this underneath type resolution.
+	if schema == tree.PublicSchema && flags.DesiredObjectKind == tree.TypeObject {
+		if alias, ok := types.PublicSchemaAliases[object]; ok {
+			if flags.RequireMutable {
+				return nil, errors.Newf("cannot use mutable descriptor of aliased type %s.%s", schema, object)
```
Comment (postamar) [inline pkg/sql/catalog/descs/collection.go:527]: Shouldn't this be an `errors.WithAssertionFailedf` instead? If not, shouldn't it have a `pgcode.WrongObjectType`?

File: pkg/sql/catalog/accessor.go [hunk 3]
```diff
@@ -13,14 +13,13 @@ package catalog
...
-	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv"
-	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 )
 
-// Accessor provides access to sql object descriptors.
+// Accessor provides access to descriptors.
+// It is implemented by the collection.
```
Comment (postamar) [inline pkg/sql/catalog/accessor.go:22]: This documentation might be worth improving a bit.

**Review by postamar — CHANGES_REQUESTED** [review]
Just a few small things. Thanks for breaking it up into small commits, this really made it easy to review.

File: pkg/sql/catalog/descs/collection.go (hunk 1 above)
Comment (ajwerner) [reply pkg/sql/catalog/descs/collection.go:2056]: Done.

File: pkg/sql/catalog/descs/collection.go (hunk 2 above)
Comment (ajwerner) [reply pkg/sql/catalog/descs/collection.go]: Done.

File: pkg/sql/catalog/accessor.go (hunk 3 above)
Comment (ajwerner) [reply pkg/sql/catalog/accessor.go]: Done

**Review by ajwerner — COMMENTED** [review]
TFTR

I believe I addressed the feedback and I rebased the schema changer catalogkv PR.

**Review by postamar — APPROVED** [review]

**Comment by ajwerner** [comment]
TFTR!

bors r=postamar

**Pull request merged**